gateway/reverse_proxy_test.go

```diff
@@ -1034,6 +1034,164 @@ func TestGraphQL_InternalDataSource(t *testing.T) {
+		spec.Proxy.ListenPath = "/tyk-subgraph-accounts-modified"
+		spec.GraphQL = apidef.GraphQLConfig{
+			Enabled:       true,
+			ExecutionMode: apidef.GraphQLExecutionModeSubgraph,
+			Version:       apidef.GraphQLConfigVersion2,
+			Schema:        gqlSubgraphSchemaAccounts,
+			Subgraph: apidef.GraphQLSubgraphConfig{
+				SDL: gqlSubgraphSDLAccounts,
+			},
+		}
+	})[0]
+	tykSubgraphBankAccounts := BuildAPI(func(spec *APISpec) {
+		spec.Name = "subgraph-bank-accounts"
+		spec.APIID = "subgraph-bank-accounts"
+		spec.Proxy.TargetURL = TestHttpAny + bankAccountSubgraphPath
+		spec.Proxy.ListenPath = "/subgraph-bank-accounts"
+		spec.GraphQL = apidef.GraphQLConfig{
+			Enabled:       true,
+			ExecutionMode: apidef.GraphQLExecutionModeSubgraph,
+			Version:       apidef.GraphQLConfigVersion2,
+			Schema:        gqlSubgraphSchemaBankAccounts,
+			Subgraph: apidef.GraphQLSubgraphConfig{
+				SDL: gqlSubgraphSDLBankAccounts,
+			},
+		}
+	})[0]
+
+	t.Run("should batch requests", func(t *testing.T) {
+		subgraphRequest := `{"query":"query($representations: [_Any!]!){_entities(representations: $representations){__typename ... on User {account {number}}}}","variables":{"representations":[{"id":"1","__typename":"User"},{"id":"2","__typename":"User"}]}}`
+		supergraph := BuildAPI(func(spec *APISpec) {
+			spec.Proxy.ListenPath = "/batched-supergraph"
+			spec.APIID = "batched-supergraph"
+			spec.GraphQL = apidef.GraphQLConfig{
+				Enabled:       true,
+				Version:       apidef.GraphQLConfigVersion2,
+				ExecutionMode: apidef.GraphQLExecutionModeSupergraph,
+				Supergraph: apidef.GraphQLSupergraphConfig{
+					Subgraphs: []apidef.GraphQLSubgraphEntity{
+						{
+							APIID: "subgraph-accounts-modified",
+							URL:   "tyk://" + tykSubgraphAccounts.Name,
+							SDL:   gqlSubgraphSDLAccounts,
+						},
+						{
+							APIID: "subgraph-bank-accounts",
+							URL:   "tyk://" + tykSubgraphBankAccounts.Name,
+							SDL:   gqlSubgraphSDLBankAccounts,
+						},
+					},
+					MergedSDL: gqlMergedSupergraphSDL,
+				},
+				Schema: gqlMergedSupergraphSDL,
+			}
+		})[0]
+		g.Gw.LoadAPI(tykSubgraphAccounts, tykSubgraphBankAccounts, supergraph)
+		timesHit := 0
+		lock := sync.Mutex{}
+		failChan := make(chan struct{})
+		g.AddDynamicHandler(bankAccountSubgraphPath, func(writer http.ResponseWriter, r *http.Request) {
+			lock.Lock()
+			timesHit++
+			lock.Unlock()
+			d, _ := io.ReadAll(r.Body)
+			defer r.Body.Close()
+			if timesHit > 1 || string(d) != subgraphRequest {
+				failChan <- struct{}{}
+				return
+			}
+		})
+
+		q := graphql.Request{
+			Query: `query Query { allUsers { id username account { number } } }`,
+		}
+		// run this in a goroutine to prevent blocking, we don't actually need the test to match body or response
+		go func() {
+			_, _ = g.Run(t, []test.TestCase{
+				{
+					Data: q, Path: "/batched-supergraph",
+				},
+			}...)
+		}()
+
+		timer := time.NewTimer(time.Second * 5)
```

Comment:
I think this approach may lead to flaky behavior under some circumstances. Instead of using locks and counters, we can utilize a `context` to check the number of executions in this case. I tried this idea using the following code:

```go
	t.Run("should batch requests", func(t *testing.T) {
		supergraph := BuildAPI(func(spec *APISpec) {
			spec.Proxy.ListenPath = "/batched-supergraph"
			spec.APIID = "batched-supergraph"
			spec.GraphQL = apidef.GraphQLConfig{
				Enabled:       true,
				Version:       apidef.GraphQLConfigVersion2,
				ExecutionMode: apidef.GraphQLExecutionModeSupergraph,
				Supergraph: apidef.GraphQLSupergraphConfig{
					Subgraphs: []apidef.GraphQLSubgraphEntity{
						{
							APIID: "subgraph-accounts-modified",
							URL:   "tyk://" + tykSubgraphAccounts.Name,
							SDL:   gqlSubgraphSDLAccounts,
						},
						{
							APIID: "subgraph-bank-accounts",
							URL:   "tyk://" + tykSubgraphBankAccounts.Name,
							SDL:   gqlSubgraphSDLBankAccounts,
						},
					},
					MergedSDL: gqlMergedSupergraphSDL,
				},
				Schema: gqlMergedSupergraphSDL,
			}
		})[0]
		g.Gw.LoadAPI(tykSubgraphAccounts, tykSubgraphBankAccounts, supergraph)
		handlerCtx, cancel := context.WithCancel(context.Background())
		g.AddDynamicHandler(bankAccountSubgraphPath, func(writer http.ResponseWriter, r *http.Request) {
			select {
			case <-handlerCtx.Done():
				assert.Fail(t, "Called twice time")
			default:
			}
			cancel()
		})

		q := graphql.Request{
			Query: `query Query { allUsers { id username account { number } } }`,
		}

		_, _ = g.Run(t, []test.TestCase{
			{
				Data: q, Path: "/batched-supergraph",
			},
		}...)
	})
```

I just removed locks and counters and used context with a cancel function. It raises a failure message if the dynamic handler ran by twice. I also removed the goroutine and timer. If I add `DisableQueryBatching: true` to the supergraph config, this test case fails with the error `Called twice time`.

Reply:
I generally avoid failing test  in goroutines other than the main goroutine (in this case the handler is called in a different goroutine), but the use of context here is much better than a timer.

Reply:
Oh it also turns out g.Run is a blocking call, so failing from the handler should be fine